Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,632 @@
+    }
+  }
+
+  @inlinable
+  internal func receiveInterceptedMetadata(_ headers: HPACKHeaders) {
+    switch self.state {
+    case .idle:
+      // Make a context to invoke the user handler with.
+      let context = GRPCAsyncServerCallContext(
+        headers: headers,
+        logger: self.context.logger,
+        userInfoRef: self.userInfoRef
+      )
+
+      // Create a source for our request stream.
+      let requestStreamSource = PassthroughMessageSource<Request, Error>()
+
+      // Create a promise to hang a callback off when the user handler completes.
+      let userHandlerPromise: EventLoopPromise<Void> = self.context.eventLoop.makePromise()
+
+      // Create a request stream from our stream source to pass to the user handler.
+      let requestStream = GRPCAsyncRequestStream(.init(consuming: requestStreamSource))
+
+      // TODO: In future use `AsyncWriter.init(maxPendingElements:maxWritesBeforeYield:delegate:)`?
+      let responseStreamWriter =
+        GRPCAsyncResponseStreamWriter(
+          wrapping: AsyncWriter(delegate: AsyncResponseStreamWriterDelegate(
+            context: context,
+            compressionIsEnabled: self.context.encoding.isEnabled,
+            send: { response, metadata in
+              self.interceptResponse(response, metadata: metadata, promise: nil)
+            },
+            finish: self.responseStreamDrained(_:)
+          ))
+        )
+
+      // Set the state to active and bundle in all the associated data.
+      self.state = .active(requestStreamSource, context, responseStreamWriter, userHandlerPromise)
+
+      // Register callback for the completion of the user handler.
+      userHandlerPromise.futureResult.whenComplete(self.userHandlerCompleted(_:))
+
+      // Send response headers back via the interceptors.
+      // TODO: In future we may want to defer this until the first response is available from the user handler which will allow the user to set the response headers via the context.
+      self.interceptors.send(.metadata([:]), promise: nil)
+
+      // Spin up a task to call the async user handler.
+      self.userHandlerTask = userHandlerPromise.completeWithTask {
+        return try await withTaskCancellationHandler {
+          do {
+            // When the user handler completes we invalidate the request stream source.
+            defer { requestStreamSource.finish() }
+            // Call the user handler.
+            try await self.userHandler(requestStream, responseStreamWriter, context)
+          } catch let status as GRPCStatus where status.isOk {
+            // The user handler throwing `GRPCStatus.ok` is considered to be invalid.
+            responseStreamWriter.asyncWriter.cancelAsynchronously()
+            throw GRPCStatus(
+              code: .unknown,
+              message: "Handler threw GRPCStatus error with code .ok"
+            )
+          } catch {
+            // TODO: Should we be instead doing `finish(throwing: error)`?
```

Comment:
Hmm, we don't currently offer a way of finishing with an error and given we only finish with a cancellation error via `cancel()` I'm not sure this is worth adding now.